### DISCUSSION TITLE
Generate rust bindings to a move contract

### DISCUSSION THREAD

**Discussion by benluelo** [aptos-labs/aptos-developer-discussions#393] [Questions]
I'm looking for something like https://alloy.rs/highlights/the-sol!-procedural-macro.html but for move code.

I've also found https://github.com/ThalaLabs/surf for typescript, but nothing for rust.

**Comment by 0x-j** [comment]
hey unfortunately there isn't a production ready rust type gen tool. I know @banool is cooking something, maybe he can share more details

**Reply by benluelo** [reply]
I hacked some codegen together: https://github.com/unionlabs/union/tree/voyager-v2/tools/move-bindgen

works quite well; example of generated code: https://github.com/unionlabs/union/blob/voyager-v2/voyager/modules/chain/movement/src/client.rs

one thing I'm confused about is how to get the abi from compiled code - currently I'm fetching the abi from on chain, which hardly seems ideal.

happy to collaborate on a more solid solution here!

**Reply by banool** [reply]
Hey hey, this is sweet! I wonder, do you have special handling for strings, options, vecs, etc. For example, when you see 0x1::string::String, treating it like a string rather than a vec of bytes. Option as an Option rather than a struct with a vec in it. etc.

I was indeed cooking something up a while ago, but my approach is... more convoluted, because I was trying to support multiple languages. In short, it does it in two phases, ABI -> GraphQL schema -> GraphQL schema to code. This is sort of nice because you only need to write the ABI -> GraphQL schema once and then you can benefit from the schema -> codegen stack for each language. The downside is GraphQL is sort of wonky for this, it is indeed query based, so lots of generators didn't work well for it. Nonetheless, if you're keen to take a look, I did get it working for Rust. Some of the logic in there for handling special types like I mentioned earlier might be helpful: https://github.com/aptos-labs/aptos-core/pull/9404. You can start with this PR, each PR has a link to the next in the comment.

As for getting the ABI from compiled code, I think actually my stack there does that, rather than uses the node API. There should be some pointers there.

We'd love to collab on this, though at this very moment we might not have time to help out too much. To push it from good to great we'd need things like special types handling, support for generating the code from different sources (node API vs compiled code, etc), support for custom derives, recursively fetching dependent modules when doing codegen (and doing this efficiently, e.g. caching), everything related to ergonomics, etc. @benluelo do you have any suggestions on the best way to track these kinds of outstanding items? I don't wanna just spam your repo with issues hahahh. I've got a lot of half cooked roadmaps for this kind of work lying around from the last 1-2 years that I'd love to consolidate somewhere.
